src/main/java/io/usethesource/vallang/IConstructor.java

```diff
@@ -25,6 +25,14 @@
  */
 public interface IConstructor extends INode {
 
+	@Override
+	default int getMatchFingerprint() {
+		return getName().hashCode() << 2 + arity();
```

Comment:
@PaulKlint @DavyLandman  I'm worried about this `<< 2` shift. In my mind, this quadruples the chance for a collision without adding much randomness to the remaining bits. We could also do this: `getName().hashCode() << 2 + 17 * arity();` to mix in the bits of the arity over the existing bits. Wouldn't that lead to a better uniform distribution?

Reply:
note to self: This happens also with INode.getMatchFingerprint

Reply:
it would only make the hash worse if the lowest 2 bits of arity are also zero right?

most jdk build-in HashMaps tend to use a mixer to protect against crappy hashes. So it's less of a problem that you might think, depending on where we are storing these fingerprints.

Reply:
Note, we could also do what `Array.hashCode` does:

```java
return (getName().hashCode() * 31 ) + arity();
```

Reply:
and maybe just make a small function:

```java
int mix(int a, int b) {
   return (a * 31) + b;
}
int mix(int a, int b, int c) {
   return ((a * 31 + b) * 31) + c;
}
```

Reply:
I've also in the [past wrote a derived version of xxhash](https://github.com/iguana-parser/iguana/blob/b6a5fd54a58ce8a9e9f3b5e8e1e839038a2e9b1c/src/org/jgll/util/hashing/XXHash.java) to mix a few integers in a way that would still give a good distribution.  We could write something like that, based on latest xxhash version.

Reply:
> it would only make the hash worse if the lowest 2 bits of arity are also zero right?

It makes it always worse for any two constructors with the same arity and a hash collision. Those collisions are 4 times more likely due to throwing away two bits of information. And since most constructors have arity 1 or 2 or 3, the chances are low that the arity actually prevents the collision.

Reply:
Multiplying instead of shifting is faster on the JVM. So if we have to shift then `*31` is just a faster way of doing that.

Reply:
> most jdk build-in HashMaps tend to use a mixer to protect against crappy hashes. So it's less of a problem that you > might think, depending on where we are storing these fingerprints.

The fingerprints are not used for implementing `hashCode` and storing in maps. They are used for the Rascal generated Java `switch` case labels for pattern matching and dynamic dispatch. Possible collisions have a direct efficiency impact that the user can not work around or fix.

Reply:
> I've also in the [past wrote a derived version of xxhash](https://github.com/iguana-> parser/iguana/blob/b6a5fd54a58ce8a9e9f3b5e8e1e839038a2e9b1c/src/org/jgll/util/hashing/XXHash.java) to mix a 

> few integers in a way that would still give a good distribution. We could write something like that, based on latest > xxhash version.

I think simply adding the arity to  the name.hashCode() is good enough. We have a very sparse table usually. All we have to make sure is that different arities with the same name end up in different buckets. That is very important since it is common to have overloaded constructors with the same name and different arities.

Reply:
> The fingerprints are not used for implementing hashCode and storing in maps. They are used for the Rascal generated Java switch case labels for pattern matching and dynamic dispatch. Possible collisions have a direct efficiency impact that the user can not work around or fix.

Ah good point. indeed.

> It makes it always worse for any two constructors with the same arity and a hash collision. Those collisions are 4 times more likely due to throwing away two bits of information. And since most constructors have arity 1 or 2 or 3, the chances are low that the arity actually prevents the collision.

if `hashCode()` and `arity()` are the same, no matter how we mix them, they would still be the same right?

> Multiplying instead of shifting is faster on the JVM. So if we have to shift then *31 is just a faster way of doing that.

in a way, `* 31` is worse than `<< 2` as 31 has a high chance of clearing off the lowest 4--5 bits. 

How about we xor the two values? the hash and the arity? that way we aren't throwing away bits of the string hash code.

Reply:
> if hashCode() and arity() are the same, no matter how we mix them, they would still be the same right?

We are trying to prevent that the fingerPrint becomes the same, by not throwing away two random bits of the hashCode of the name, and by making sure that at least nodes with different arities have different fingerprints.

Reply:
> How about we xor the two values? the hash and the arity? that way we aren't throwing away bits of the string hash code.

`xor` is much slower than multiplication and addition, while they have almost the same effect on the uniformity. Also, xor must usually also be combined with a shift or `*31` extra, since the arities are always low.

Reply:
ok, lets leave it as is